kvserver/load: add node capacity provider

wenyihu6 opened the pull request:
**kvserver/load: add node capacity provider**

This commit introduces NodeCapacityProvider, which monitors runtime cpu usage,
capacity and also aggregates cpu usage across all stores. It is not yet
integrated with other components. Future commits will hook it into each store to
populate the node capacity field in the store descriptor.

Epic: none
Release none

---

**kvserver: plumb nodeCapacityProvider**

Previously, we added a NodeCapacityProvider field to StoreGossip but left it nil
in production code In addition, we had a TODO in
https://github.com/cockroachdb/cockroach/commit/1b6a91fcda9315683d223ed5f00ca29728fdbdb6
to find a more appropriate home for node capacity provider, since the runtime
load monitor is a node-level concept and was being passed to NewStores. During
implementation, we identified a better home for it. This commit moves
NodeCapacityProvider to be a node level concept as part of StoreConfig and
passes it to each store for use when constructing the store descriptor. Note
that runtime load monitoring currently lacks caching and just computes
periodically. We may consider this as a future improvement.

Epic: none
Release note: none

wenyihu6 committed:
kvserver: plumb nodeCapacityProvider

Previously, we added a NodeCapacityProvider field to StoreGossip but left it nil
in production code In addition, we had a TODO in
https://github.com/cockroachdb/cockroach/commit/1b6a91fcda9315683d223ed5f00ca29728fdbdb6
to find a more appropriate home for node capacity provider, since the runtime
load monitor is a node-level concept and was being passed to NewStores. During
implementation, we identified a better home for it. This commit moves
NodeCapacityProvider to be a node level concept as part of StoreConfig and
passes it to each store for use when constructing the store descriptor. Note
that runtime load monitoring currently lacks caching and just computes
periodically. We may consider this as a future improvement.

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/150275)
<!-- Reviewable:end -->

wenyihu6 committed:
kvserver/load: add NodeCapacityProvider

Previously, we added a NodeCapacityProvider field to StoreGossip but left it nil
in production code. In addition, we had a TODO in
https://github.com/cockroachdb/cockroach/commit/1b6a91fcda9315683d223ed5f00ca29728fdbdb6
to find a more appropriate home for node capacity provider, since the runtime
load monitor is a node-level concept and was being passed to NewStores. During
implementation, we identified a better home for it. This commit moves
NodeCapacityProvider to be a node level concept as part of StoreConfig and
passes it to each store for use when constructing the store descriptor. Note
that runtime load monitoring currently lacks caching, and we may add it later if
it proves too expensive.

Epic: none
Release note: none

tbg reviewed (changes requested):
I left a bunch of comments,  but this looks really good!
Please read all of my comments before acting on them, as some of them don't pan out given the final state of the PR (need testing knobs, for example). I do think it makes sense to unify the runtime load monitor and the node capacity provider, just to keep things tight and to avoid redundant testing.

> Reviewed 5 of 5 files at r1, 4 of 4 files at r2, 10 of 10 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-:-OVHG3HL7cOVH2d0DYVG:bmm5vh2)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`pkg/kv/kvserver/load/runtime_load_monitor_test.go` line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVH6yEX4XAmAXPqzsdj:-OVH6yEX4XAmAXPqzsdk:bxz47yb) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a3ca884802776a965e8a8b6116b021bff3156a8/pkg/kv/kvserver/load/runtime_load_monitor_test.go#L23)):*
> ```Go
> // RuntimeLoadMonitor.
> func TestRuntimeLoadMonitor(t *testing.T) {
> 	stopper := stop.NewStopper()
> ```

Should have a stopper in each test case, or goroutines from one test will live into the other.
___
*[`pkg/kv/kvserver/load/runtime_load_monitor_test.go` line 33 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVH6mNp1ACYyqnpLRs_:-OVH6mNp1ACYyqnpLRsa:b-anlj1n) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a3ca884802776a965e8a8b6116b021bff3156a8/pkg/kv/kvserver/load/runtime_load_monitor_test.go#L33)):*
> ```Go
> 		assert.IsType(t, &RunTimeLoadStats{}, stats)
> 
> 		assert.GreaterOrEqual(t, stats.CPUUsageNanoPerSec, int64(0))
> ```

How could this ever fail? The other test case eventually wants it to be nonzero.

Looks like you  could get away with just one test case here - there's a lot of duplication.
___
*[`pkg/kv/kvserver/load/runtime_load_monitor_test.go` line 34 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVH6cxR-ER57Z7tZy_k:-OVH6cxR-ER57Z7tZy_l:b376bw5) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a3ca884802776a965e8a8b6116b021bff3156a8/pkg/kv/kvserver/load/runtime_load_monitor_test.go#L34)):*
> ```Go
> 
> 		assert.GreaterOrEqual(t, stats.CPUUsageNanoPerSec, int64(0))
> 		assert.GreaterOrEqual(t, stats.CPUCapacityNanoPerSec, int64(0))
> ```

shouldn't this be >0?
___
*[`pkg/kv/kvserver/load/runtime_load_monitor_test.go` line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVH7DaX2Fkj928wOFLP:-OVH7DaX2Fkj928wOFLQ:betikvt) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a3ca884802776a965e8a8b6116b021bff3156a8/pkg/kv/kvserver/load/runtime_load_monitor_test.go#L57)):*
> ```Go
> 		cancel()
> 
> 		// Make sure the monitor does not return nil after cancellation.
> ```

Just have it return a value? I was wondering why that method returned a pointer. You could then still check that it still returns a nonzero capacity.
___
*[`pkg/kv/kvserver/load/runtime_load_monitor.go` line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVH4--l6_S8Euvf-QBL:-OVH4--l6_S8Euvf-QBM:bued5f5) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a3ca884802776a965e8a8b6116b021bff3156a8/pkg/kv/kvserver/load/runtime_load_monitor.go#L23)):*
> ```Go
> // weighted more heavily, and data from the past 20 seconds contributes
> // to the average.
> const defaultMovingAverageAge = 20
> ```

Let's move this into `newRuntimeLoadMonitor`, which is the only place where it's used.

The comment isn't really accurate: all past samples influence the EWMA value. It's just that samples earlier than the 20th-to-most-recent barely make a dent any more due to exponential decay.

Also, it's a bit inconsistent that the sampling interval is passed into `newRuntimeLoadMonitor` but this here isn't. My suggestion would be to make all of the consts local to `NewNodeCapacityProvider` (incl defaultCPUUsageRefreshInterval and its sibling), and to pass the decay factor into `newRuntimeLoadMonitor` as well. (No need to patch up existing commits, just add new ones).
___
*[`pkg/kv/kvserver/load/runtime_load_monitor.go` line 35 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVH2dkV1Z1BVlKu75DJ:-OVH2dkWAKcwFJv9qA7Q:byt3dm2) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a3ca884802776a965e8a8b6116b021bff3156a8/pkg/kv/kvserver/load/runtime_load_monitor.go#L35)):*
> ```Go
> 
> // RuntimeLoadMonitor is the monitor for the runtime load. It measures the cpu
> // usage and capacity of the node directly using the status package.
> ```

"using the status package" isn't that helpful. Maybe a more succinct yet also more descriptive overall comment would be:

// RuntimeLoadMonitor polls the CPU utilization and maintains a moving average.
___
*[`pkg/kv/kvserver/load/runtime_load_monitor.go` line 85 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVH5MNqCZisAGnBaQZ4:-OVH5MNqCZisAGnBaQZ5:bo25mm) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a3ca884802776a965e8a8b6116b021bff3156a8/pkg/kv/kvserver/load/runtime_load_monitor.go#L85)):*
> ```Go
> 	defer m.mu.Unlock()
> 	userTime, sysTime, err := status.GetProcCPUTime(ctx)
> 	if buildutil.CrdbTestBuild && err != nil {
> ```

In production, this should at least log a warning the first time around. We also need `GetCPUStats()` to make sense even in this case - it needs to somehow reflect that we have no meaningful way to determine the CPU usage, probably by returning a zero usage AND capacity.
___
*[`pkg/kv/kvserver/store_test.go` line 41 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVHBgul2Bsw99FhGiik:-OVHBgul2Bsw99FhGiil:b-d57vix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1332e095320fd04501eeb006b4112c921192e1bc/pkg/kv/kvserver/store_test.go#L41)):*
> ```Go
> 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvstorage"
> 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/liveness/livenesspb"
> 	loadmonitor "github.com/cockroachdb/cockroach/pkg/kv/kvserver/load"
> ```

what's the reason for the custom import name here and in some other file? `load` is already taken by something?
___
*[`pkg/kv/kvserver/store_test.go` line 4234 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVHD3fyBqJyFvIptuBD:-OVHD3fyBqJyFvIptuBE:bpwn13o) ([raw file](https://github.com/cockroachdb/cockroach/blob/1332e095320fd04501eeb006b4112c921192e1bc/pkg/kv/kvserver/store_test.go#L4234)):*
> ```Go
> 	serverArgs := base.TestServerArgs{
> 		Knobs: base.TestingKnobs{
> 			NodeCapacityProviderKnobs: &loadmonitor.NodeCapacityProviderTestingKnobs{
> ```

I see, so you do need some proper testing knobs - makes sense, please take my earlier comment with a grain of salt. I still think we can consolidate between the runtime stats monitor and the node capacity monitor, and apply the testing knobs as high up as possible.
___
*[`pkg/kv/kvserver/store_test.go` line 4259 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVHAbtK90i_H6Bk0oqd:-OVHAbtK90i_H6Bk0oqe:b-f8ltyi) ([raw file](https://github.com/cockroachdb/cockroach/blob/1332e095320fd04501eeb006b4112c921192e1bc/pkg/kv/kvserver/store_test.go#L4259)):*
> ```Go
> 		require.Equal(t, int32(numStoresPerNode), nc.NumStores)
> 		if nc.NodeCPURateUsage == 0 || nc.NodeCPURateCapacity == 0 || nc.StoresCPURate == 0 {
> 			return errors.New("CPU usage or capacity is 0")
> ```

include the values in the error, so that we can see which one is the offending one.

This relies on there being some background work in the system; I wouldn't be too surprised if this were flaky but let's leave it as is (i.e. very aggressive) for now. The retry loop probably makes this always pass.
___
*[`pkg/server/status/runtime.go` line 1260 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVH6WY9ED5cu0YcMXFu:-OVH6WY9ED5cu0YcMXFv:b-y47lu7) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a3ca884802776a965e8a8b6116b021bff3156a8/pkg/server/status/runtime.go#L1260)):*
> ```Go
> }
> 
> // getCPUCapacity returns the number of logical CPU processors available for
> ```

GetCPUCapacity
___
*[`pkg/kv/kvserver/load/node_capacity_provider.go` line 26 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVH8n4GA9H4ziE88Bma:-OVH8n4H2JGg_35lqyYk:b-9e1cef) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed505faf9a1555aaea3694a0fe2be5ae8f64482a/pkg/kv/kvserver/load/node_capacity_provider.go#L26)):*
> ```Go
> // NodeCapacityProvider monitors and reports node-level CPU capacity and usage
> // by aggregating stores and runtime stats.
> type NodeCapacityProvider struct {
> ```

Looking at this, does the division between RuntimeLoadMonitor (RLM) and NodeCapacityProvider (NCP) make sense? I have a few gripes with this - the NCA just starts the wrapped RLM, and any test against the RLM can be written against the NCP directly (just provide a mock StoresStatsAggregator). It seems simpler to merge the RLM completely into the NCP. Then we have one struct that is in charge of measuring CPU capacity, CPU utilization by node, and it can pull in the CPU utilization by store as well. That seems like a sensible set of things it can do.
___
*[`pkg/kv/kvserver/load/node_capacity_provider.go` line 47 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVH8BwW4jaTyHioJgzk:-OVH8BwW4jaTyHioJgzl:b30gos6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed505faf9a1555aaea3694a0fe2be5ae8f64482a/pkg/kv/kvserver/load/node_capacity_provider.go#L47)):*
> ```Go
> // allows customizing refresh intervals for testing.
> func NewNodeCapacityProvider(
> 	stopper *stop.Stopper, stores StoresStatsAggregator, knobs *NodeCapacityProviderTestingKnobs,
> ```

Instead of passing knobs that override the defaults, can you pass in the options directly? Then there is no need for knobs and the code should be clearer.
___
*[`pkg/kv/kvserver/load/node_capacity_provider.go` line 63 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVH9Rst5oF6_3J9Ek-e:-OVH9Rst5oF6_3J9Ek-f:b-sjqby1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed505faf9a1555aaea3694a0fe2be5ae8f64482a/pkg/kv/kvserver/load/node_capacity_provider.go#L63)):*
> ```Go
> // Run starts the background monitoring of CPU metrics.
> func (n *NodeCapacityProvider) Run(ctx context.Context) {
> 	n.runtimeLoadMonitor.Run(ctx)
> ```

I forgot to mention it in the last commit - it may make sense to poll the capacity synchronously as part of `Run`, so that we don't ever return a zero capacity for immediate subsequent calls to retrieve the stats. (We'll return a zero usage, but that's hard to avoid and should be ok, though we could add a "ready" period at some point if we find that newly started nodes are receiving replicas they shouldn't get -  I don't think the system is fast enough for this to become an issue).
___
*[`pkg/kv/kvserver/load/node_capacity_provider.go` line 68 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVH9yTb59ZzYEdB66YD:-OVH9yTb59ZzYEdB66YE:b-9o8vqn) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed505faf9a1555aaea3694a0fe2be5ae8f64482a/pkg/kv/kvserver/load/node_capacity_provider.go#L68)):*
> ```Go
> // GetNodeCapacity returns the current CPU capacity and usage metrics for the
> // node.
> func (n *NodeCapacityProvider) GetNodeCapacity(useCached bool) roachpb.NodeCapacity {
> ```

Mildly awkward that `useCached` only applies to the stores stats. Add a comment and maybe a TODO to revisit once this is all hooked up. Perhaps the callers that specify `false` will also want to slip in an extra update for the node CPU, though I hope this wouldn't be necessary given that we poll every second.
___
*[`pkg/kv/kvserver/stores.go` line 339 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVHC3tN2M6IhiC2n859:-OVHC3tN2M6IhiC2n85A:b-3tcgqr) ([raw file](https://github.com/cockroachdb/cockroach/blob/1332e095320fd04501eeb006b4112c921192e1bc/pkg/kv/kvserver/stores.go#L339)):*
> ```Go
> }
> 
> // TODO(wenyihu6): Implment properly for production code. This now just returns
> ```

Remove this TODO.


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
TFTR! It does look a lot cleaner now. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-:-OVJORW79XdetnpDsGl0:b-ybk63v)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[`pkg/kv/kvserver/stores.go` line 339 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVHC3tN2M6IhiC2n859:-OVJOdYRAvjL3dpojra3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1332e095320fd04501eeb006b4112c921192e1bc/pkg/kv/kvserver/stores.go#L339)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Remove this TODO.
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/load/node_capacity_provider.go` line 26 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVH8n4GA9H4ziE88Bma:-OVJOfpCA7gmccrcPmwb:ba9k66m) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed505faf9a1555aaea3694a0fe2be5ae8f64482a/pkg/kv/kvserver/load/node_capacity_provider.go#L26)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Looking at this, does the division between RuntimeLoadMonitor (RLM) and NodeCapacityProvider (NCP) make sense? I have a few gripes with this - the NCA just starts the wrapped RLM, and any test against the RLM can be written against the NCP directly (just provide a mock StoresStatsAggregator). It seems simpler to merge the RLM completely into the NCP. Then we have one struct that is in charge of measuring CPU capacity, CPU utilization by node, and it can pull in the CPU utilization by store as well. That seems like a sensible set of things it can do.
</blockquote></details>

Agreed. I still kept a separate struct for runtimeLoadMonitor, but I’ve now moved the `runtimeLoadMonitor` entirely into `nodeCapacityProvider` as a private helper struct. Does this align with what you had in mind? 
___
*[`pkg/kv/kvserver/load/node_capacity_provider.go` line 47 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVH8BwW4jaTyHioJgzk:-OVJP2124f1IZlwe2Ro4:b-5kuez5) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed505faf9a1555aaea3694a0fe2be5ae8f64482a/pkg/kv/kvserver/load/node_capacity_provider.go#L47)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Instead of passing knobs that override the defaults, can you pass in the options directly? Then there is no need for knobs and the code should be clearer.
</blockquote></details>

This comment looks stale now given what you mentioned elsewhere that testing knob makes sense in real cluster setups. Lmk if I'm misunderstanding anything tho. 
___
*[`pkg/kv/kvserver/load/node_capacity_provider.go` line 63 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVH9Rst5oF6_3J9Ek-e:-OVJPIipD8Ftv3QK5ar-:b2ydt7v) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed505faf9a1555aaea3694a0fe2be5ae8f64482a/pkg/kv/kvserver/load/node_capacity_provider.go#L63)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I forgot to mention it in the last commit - it may make sense to poll the capacity synchronously as part of `Run`, so that we don't ever return a zero capacity for immediate subsequent calls to retrieve the stats. (We'll return a zero usage, but that's hard to avoid and should be ok, though we could add a "ready" period at some point if we find that newly started nodes are receiving replicas they shouldn't get -  I don't think the system is fast enough for this to become an issue).
</blockquote></details>

Agreed. I'm now calling `recordCPUCapacity` once during `NewNodeCapacityProvider` initialization. I put it there instead of in `Run` since I didn’t want to rely on `Run` being called before `GetNodeCapacity`.
___
*[`pkg/kv/kvserver/load/node_capacity_provider.go` line 68 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVH9yTb59ZzYEdB66YD:-OVJPjtE74Tl-HlRGYYi:b-xhwbj8) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed505faf9a1555aaea3694a0fe2be5ae8f64482a/pkg/kv/kvserver/load/node_capacity_provider.go#L68)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Mildly awkward that `useCached` only applies to the stores stats. Add a comment and maybe a TODO to revisit once this is all hooked up. Perhaps the callers that specify `false` will also want to slip in an extra update for the node CPU, though I hope this wouldn't be necessary given that we poll every second.
</blockquote></details>

Added a TODO. 
___
*[`pkg/server/status/runtime.go` line 1260 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVH6WY9ED5cu0YcMXFu:-OVJSz6eDvlxmavRkha3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a3ca884802776a965e8a8b6116b021bff3156a8/pkg/server/status/runtime.go#L1260)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

GetCPUCapacity
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/store_test.go` line 41 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVHBgul2Bsw99FhGiik:-OVJTECL0gToCN1_KFIM:b-8xvjs9) ([raw file](https://github.com/cockroachdb/cockroach/blob/1332e095320fd04501eeb006b4112c921192e1bc/pkg/kv/kvserver/store_test.go#L41)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

what's the reason for the custom import name here and in some other file? `load` is already taken by something?
</blockquote></details>

Not for this file, removed.
___
*[`pkg/kv/kvserver/store_test.go` line 4234 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVHD3fyBqJyFvIptuBD:-OVJTOCiC_pRJ58Qlad0:b-55oevn) ([raw file](https://github.com/cockroachdb/cockroach/blob/1332e095320fd04501eeb006b4112c921192e1bc/pkg/kv/kvserver/store_test.go#L4234)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I see, so you do need some proper testing knobs - makes sense, please take my earlier comment with a grain of salt. I still think we can consolidate between the runtime stats monitor and the node capacity monitor, and apply the testing knobs as high up as possible.
</blockquote></details>

Updated node capacity monitor and runtime stats tgtr. 
___
*[`pkg/kv/kvserver/store_test.go` line 4259 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVHAbtK90i_H6Bk0oqd:-OVJXsYf4CjeG65JKPrv:b-k7vswl) ([raw file](https://github.com/cockroachdb/cockroach/blob/1332e095320fd04501eeb006b4112c921192e1bc/pkg/kv/kvserver/store_test.go#L4259)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

include the values in the error, so that we can see which one is the offending one.

This relies on there being some background work in the system; I wouldn't be too surprised if this were flaky but let's leave it as is (i.e. very aggressive) for now. The retry loop probably makes this always pass.
</blockquote></details>

Added specific values. 

re: flaky, I will keep it in mind and check for CI failures once merged. 
___
*[`pkg/kv/kvserver/load/runtime_load_monitor.go` line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVH4--l6_S8Euvf-QBL:-OVJYNRq2FwmszGgiCEP:b-cgca45) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a3ca884802776a965e8a8b6116b021bff3156a8/pkg/kv/kvserver/load/runtime_load_monitor.go#L23)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Let's move this into `newRuntimeLoadMonitor`, which is the only place where it's used.

The comment isn't really accurate: all past samples influence the EWMA value. It's just that samples earlier than the 20th-to-most-recent barely make a dent any more due to exponential decay.

Also, it's a bit inconsistent that the sampling interval is passed into `newRuntimeLoadMonitor` but this here isn't. My suggestion would be to make all of the consts local to `NewNodeCapacityProvider` (incl defaultCPUUsageRefreshInterval and its sibling), and to pass the decay factor into `newRuntimeLoadMonitor` as well. (No need to patch up existing commits, just add new ones).
</blockquote></details>

I ended up just getting rid of `newRuntimeLoadMonitor` and making `NewNodeCapacityProvider` construct runtimeLoadMonitor. 
___
*[`pkg/kv/kvserver/load/runtime_load_monitor.go` line 35 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVH2dkV1Z1BVlKu75DJ:-OVJYII0EBFrK9jwbXBW:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a3ca884802776a965e8a8b6116b021bff3156a8/pkg/kv/kvserver/load/runtime_load_monitor.go#L35)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

"using the status package" isn't that helpful. Maybe a more succinct yet also more descriptive overall comment would be:

// RuntimeLoadMonitor polls the CPU utilization and maintains a moving average.
</blockquote></details>

Done. 
___
*[`pkg/kv/kvserver/load/runtime_load_monitor.go` line 85 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVH5MNqCZisAGnBaQZ4:-OVJYc5K52CWcfxHNvfy:bsbh44) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a3ca884802776a965e8a8b6116b021bff3156a8/pkg/kv/kvserver/load/runtime_load_monitor.go#L85)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

In production, this should at least log a warning the first time around. We also need `GetCPUStats()` to make sense even in this case - it needs to somehow reflect that we have no meaningful way to determine the CPU usage, probably by returning a zero usage AND capacity.
</blockquote></details>

I wasn't sure of a clean way to handle this without adding state to the struct, and it's unclear how frequently this error occurs. For now, I'm logging a warning and recording 0. If this error case persists, I'd expect the CPU usage to trend toward 0 as we measure another 20 measurements (20s). We can follow up with more targeted error handling in future PRs if needed.
___
*[`pkg/kv/kvserver/load/runtime_load_monitor_test.go` line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVH6yEX4XAmAXPqzsdj:-OVJYAPoBulfwvFfgYFf:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a3ca884802776a965e8a8b6116b021bff3156a8/pkg/kv/kvserver/load/runtime_load_monitor_test.go#L23)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Should have a stopper in each test case, or goroutines from one test will live into the other.
</blockquote></details>

Removed.
___
*[`pkg/kv/kvserver/load/runtime_load_monitor_test.go` line 33 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVH6mNp1ACYyqnpLRs_:-OVJY8gm8W_yohkucVHa:b4rhtra) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a3ca884802776a965e8a8b6116b021bff3156a8/pkg/kv/kvserver/load/runtime_load_monitor_test.go#L33)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

How could this ever fail? The other test case eventually wants it to be nonzero.

Looks like you  could get away with just one test case here - there's a lot of duplication.
</blockquote></details>

Same as above, removed.
___
*[`pkg/kv/kvserver/load/runtime_load_monitor_test.go` line 34 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVH6cxR-ER57Z7tZy_k:-OVJY1Xl6JUB_7TztvoJ:b-a4m0g) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a3ca884802776a965e8a8b6116b021bff3156a8/pkg/kv/kvserver/load/runtime_load_monitor_test.go#L34)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

shouldn't this be >0?
</blockquote></details>

Removed this file given we are moving run time load monitor to node capacity provider. 
___
*[`pkg/kv/kvserver/load/runtime_load_monitor_test.go` line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVH7DaX2Fkj928wOFLP:-OVJYCEK4Xhmp9IL37XS:bjbkz1c) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a3ca884802776a965e8a8b6116b021bff3156a8/pkg/kv/kvserver/load/runtime_load_monitor_test.go#L57)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Just have it return a value? I was wondering why that method returned a pointer. You could then still check that it still returns a nonzero capacity.
</blockquote></details>

This was a mistake. We should return a value. Done. 


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
Reviewable fed me a chopped up version of the complete diffs again, please either teach me tricks that make this easier or prefer an append-only history :sweat_smile:
Looks good, thank you! Remainder are nits, I feel somewhat strongly on the testing knob thing but if it gives you more trouble than I think it should, it should not hold up this PR.

> Reviewed 16 of 16 files at r4, 12 of 12 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-:-OVJgX3W2kL8nl_wN6p1:b2brg58)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`pkg/kv/kvserver/load/node_capacity_provider.go` line 47 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVH8BwW4jaTyHioJgzk:-OVJdwFM-e2UAUxBghZf:b-gbe4di) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed505faf9a1555aaea3694a0fe2be5ae8f64482a/pkg/kv/kvserver/load/node_capacity_provider.go#L47)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

This comment looks stale now given what you mentioned elsewhere that testing knob makes sense in real cluster setups. Lmk if I'm misunderstanding anything tho.
</blockquote></details>

I was thinking to terminate the testing knob higher up. Namely, `NewNodeCapacityProvider` should take its configuration properly, and whoever the caller is (or the caller's caller, etc, as high up as we can go) handles the knobs. That way, knobs don't go to places they don't need to be in, and defaults are set closer to where the server is configured, which is usually preferable.
___
*[`pkg/kv/kvserver/load/runtime_load_monitor.go` line 85 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVH5MNqCZisAGnBaQZ4:-OVJgR4bCualtzPWr4Xn:b-jmneww) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a3ca884802776a965e8a8b6116b021bff3156a8/pkg/kv/kvserver/load/runtime_load_monitor.go#L85)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

I wasn't sure of a clean way to handle this without adding state to the struct, and it's unclear how frequently this error occurs. For now, I'm logging a warning and recording 0. If this error case persists, I'd expect the CPU usage to trend toward 0 as we measure another 20 measurements (20s). We can follow up with more targeted error handling in future PRs if needed.
</blockquote></details>

As per the other command, let's shelve this now and punt it to productionization phase.
___
*[`pkg/kv/kvserver/load/node_capacity_provider.go` line 104 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVJePyU4fWKpNEqDK3n:-OVJePyU4fWKpNEqDK3o:b-o3haey) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c06e1e526eadaeb706dbca6e3e1881a3dffa637/pkg/kv/kvserver/load/node_capacity_provider.go#L104)):*
> ```Go
> 		NumStores:           numStores,
> 		NodeCPURateCapacity: cpuCapacityNanoPerSec,
> 		NodeCPURateUsage:    cpuUsageNanoPerSec,
> ```

Not for this PR but perhaps a line item in our spreadsheet: should this and the capacity also become metrics? Or is the stores rate already published per store as a metric? In that case, we only have to talk about the CPU capacity.
___
*[`pkg/kv/kvserver/load/node_capacity_provider.go` line 121 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVJejSk5KwdRx7kHP48:-OVJejSk5KwdRx7kHP49:bhhhv64) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c06e1e526eadaeb706dbca6e3e1881a3dffa637/pkg/kv/kvserver/load/node_capacity_provider.go#L121)):*
> ```Go
> 		lastTotalUsageNanos float64
> 		// usageEWMA maintains a moving average of cpu usage in nanoseconds,
> 		// obtained by polling incremental stats from status.GetProcCPUTime.
> ```

Technically speaking, this receives the delta between subsequent polls, right? Might be worth clarifying.
___
*[`pkg/kv/kvserver/load/node_capacity_provider.go` line 146 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVJfR6P6mThIW5CXvuQ:-OVJfR6P6mThIW5CXvuR:b1s4n6i) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c06e1e526eadaeb706dbca6e3e1881a3dffa637/pkg/kv/kvserver/load/node_capacity_provider.go#L146)):*
> ```Go
> 	m.mu.Lock()
> 	defer m.mu.Unlock()
> 	userTime, sysTime, err := status.GetProcCPUTime(ctx)
> ```

if you wanted to be extra explicit: userMillis, sysMillis
___
*[`pkg/kv/kvserver/load/node_capacity_provider.go` line 151 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVJf7AG0rHRMRjc2Ehi:-OVJf7AG0rHRMRjc2Ehj:bzc4zs0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c06e1e526eadaeb706dbca6e3e1881a3dffa637/pkg/kv/kvserver/load/node_capacity_provider.go#L151)):*
> ```Go
> 			panic(err)
> 		}
> 		log.Warningf(ctx, "failed to get cpu usage: %v", err)
> ```

Is this good enough? If this happens in prod, we're in a probably untested regime.
Could you file a line item in our spreadsheet so that we revisit during productionization?
___
*[`pkg/kv/kvserver/load/node_capacity_provider.go` line 168 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVJfZpiC5-fpkNM0lnq:-OVJfZpiC5-fpkNM0lnr:bn0pkye) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c06e1e526eadaeb706dbca6e3e1881a3dffa637/pkg/kv/kvserver/load/node_capacity_provider.go#L168)):*
> ```Go
> 	m.mu.logicalCPUsPerSec = int64(status.GetCPUCapacity())
> 	if m.mu.logicalCPUsPerSec == 0 {
> 		panic("programming error: cpu capacity is 0")
> ```

under test build only, right?
___
*[`pkg/kv/kvserver/load/node_capacity_provider_test.go` line 49 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVJg1RhAGRSzMPB-oOI:-OVJg1RhAGRSzMPB-oOJ:biw3n3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c06e1e526eadaeb706dbca6e3e1881a3dffa637/pkg/kv/kvserver/load/node_capacity_provider_test.go#L49)):*
> ```Go
> 	activeProvider.Run(context.Background())
> 
> 	ctx, cancel := context.WithCancel(context.Background())
> ```

This is just a nit, but there seems to be no strong reason to define this here. In fact, you could just cancel the active provider before testing the cancellation case.
___
*[`pkg/kv/kvserver/load/node_capacity_provider_test.go` line 60 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVJfr1U73-s0uNNFbrN:-OVJfr1U73-s0uNNFbrO:b-cmu8mn) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c06e1e526eadaeb706dbca6e3e1881a3dffa637/pkg/kv/kvserver/load/node_capacity_provider_test.go#L60)):*
> ```Go
> 	testutils.SucceedsSoon(t, func() error {
> 		nc := activeProvider.GetNodeCapacity(false)
> 		require.NotNil(t, nc)
> ```

Is this a pointer? I thought it would be a struct.


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
Sg, I will follow up with a new PR on the testing knob thing. Does Reviewable get confused if I squash or fix up between recent commits I'm appending to address feedback and you haven't seen yet?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-:-OVJw2V78WOptDC_ssF2:b8h2m5e)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[`pkg/kv/kvserver/load/node_capacity_provider.go` line 47 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVH8BwW4jaTyHioJgzk:-OVJx3wj3_RXMVagTUSF:buor9cf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed505faf9a1555aaea3694a0fe2be5ae8f64482a/pkg/kv/kvserver/load/node_capacity_provider.go#L47)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I was thinking to terminate the testing knob higher up. Namely, `NewNodeCapacityProvider` should take its configuration properly, and whoever the caller is (or the caller's caller, etc, as high up as we can go) handles the knobs. That way, knobs don't go to places they don't need to be in, and defaults are set closer to where the server is configured, which is usually preferable.
</blockquote></details>

Gotcha, I will put up a pr to follow up on this. 
___
*[`pkg/kv/kvserver/load/node_capacity_provider.go` line 104 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVJePyU4fWKpNEqDK3n:-OVJxRh2BvOxsGYvSxXx:b-b252tq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c06e1e526eadaeb706dbca6e3e1881a3dffa637/pkg/kv/kvserver/load/node_capacity_provider.go#L104)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Not for this PR but perhaps a line item in our spreadsheet: should this and the capacity also become metrics? Or is the stores rate already published per store as a metric? In that case, we only have to talk about the CPU capacity.
</blockquote></details>

Added a new item on our spreadsheet. 
___
*[`pkg/kv/kvserver/load/node_capacity_provider.go` line 121 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVJejSk5KwdRx7kHP48:-OVJxnekCT-WgM5fPOW6:b-48vilg) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c06e1e526eadaeb706dbca6e3e1881a3dffa637/pkg/kv/kvserver/load/node_capacity_provider.go#L121)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Technically speaking, this receives the delta between subsequent polls, right? Might be worth clarifying.
</blockquote></details>

Good call, done.
___
*[`pkg/kv/kvserver/load/node_capacity_provider.go` line 146 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVJfR6P6mThIW5CXvuQ:-OVJxpt19sUuKUH6WIvn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c06e1e526eadaeb706dbca6e3e1881a3dffa637/pkg/kv/kvserver/load/node_capacity_provider.go#L146)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

if you wanted to be extra explicit: userMillis, sysMillis
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/load/node_capacity_provider.go` line 151 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVJf7AG0rHRMRjc2Ehi:-OVJxtJ_26n6wrFCtoEb:b-epibbd) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c06e1e526eadaeb706dbca6e3e1881a3dffa637/pkg/kv/kvserver/load/node_capacity_provider.go#L151)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Is this good enough? If this happens in prod, we're in a probably untested regime.
Could you file a line item in our spreadsheet so that we revisit during productionization?
</blockquote></details>

Added a todo in the code and in our spreadsheet. 
___
*[`pkg/kv/kvserver/load/node_capacity_provider.go` line 168 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVJfZpiC5-fpkNM0lnq:-OVJxwMs4yzLG_149c7B:b40fu9e) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c06e1e526eadaeb706dbca6e3e1881a3dffa637/pkg/kv/kvserver/load/node_capacity_provider.go#L168)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

under test build only, right?
</blockquote></details>

Added the flag. 
___
*[`pkg/kv/kvserver/load/node_capacity_provider_test.go` line 49 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVJg1RhAGRSzMPB-oOI:-OVJxxvw8xnnmMzKCZKy:b-jrq2wn) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c06e1e526eadaeb706dbca6e3e1881a3dffa637/pkg/kv/kvserver/load/node_capacity_provider_test.go#L49)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This is just a nit, but there seems to be no strong reason to define this here. In fact, you could just cancel the active provider before testing the cancellation case.
</blockquote></details>

Changed it to just use one provider. 
___
*[`pkg/kv/kvserver/load/node_capacity_provider_test.go` line 60 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVJfr1U73-s0uNNFbrN:-OVJy1gL-l0xrpEbgeCC:b45zmv0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c06e1e526eadaeb706dbca6e3e1881a3dffa637/pkg/kv/kvserver/load/node_capacity_provider_test.go#L60)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Is this a pointer? I thought it would be a struct.
</blockquote></details>

Good call, this is stale. It should be just a struct. 
___
*[`pkg/kv/kvserver/load/runtime_load_monitor.go` line 85 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-OVH5MNqCZisAGnBaQZ4:-OVJyEVS6vsSsAQdhQKt:b-5lxy3m) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a3ca884802776a965e8a8b6116b021bff3156a8/pkg/kv/kvserver/load/runtime_load_monitor.go#L85)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

As per the other command, let's shelve this now and punt it to productionization phase.
</blockquote></details>

Added to spreadsheet. 


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
It will try to diff "the old commit" and the "new version of the old commit" (which as far as git are concerned have no relationship with each other) but its heuristics don't always work. I'll try to pay close attention to where it goes wrong next time I notice it.
What do you do with a pure Git review? You basically have to review again from scratch? I think we should avoid that, since it leads to skimming. Ideally reviewers can see the clean deltas since their last reviews. The only way I know how to achieve that is using an append-only history (that optionally gets squashed before the merge).

PR looks good, :ship:!

> Reviewed 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150275#-:-OVJyleM9_SC4_yOgkh8:bl93q4i)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)



<!-- Sent from Reviewable.io -->

wenyihu6 commented:
Things are already flaky based on CI:  `require.GreaterOrEqual(t, nc.NodeCPURateCapacity, nc.NodeCPURateUsage)` failed with 
```
pkg/testutils/soon.go:38
pkg/kv/kvserver/store_test.go:4253
Error:        "16000000000" is not greater than or equal to "24230879442"
```

I will try to stress it and see what is going on.

wenyihu6 commented:
Wasn't able to repro locally. I'm going to remove this assertion and merge this in for now and add a todo to investigate. Perhaps the assumption of cpu capacity >= cpu usage doesn't hold during temporary bursts.

wenyihu6 commented:
The last push fixed some build failures and removed an assertion in tests as noted earlier.

tbg reviewed (approved):
A few things I don't understand.

- why did the test move to `load_test`?
- why is there an `-ldl` linker flag now?
- could you say a word about the plumbing changes? I see that `NewStoreGossip` used to take the provider, now it doesn't, at least based on [this force push](https://github.com/cockroachdb/cockroach/compare/a29c42e356ff701482c57e2569fc8ee53d7a6499..bb417f5149eba5d93c2e84896d5bc6b53f23204c)
- That same diff also removes `simNodeCapacityProvider`, I'm assuming it was unused as a result of the previous merging of node capacity provider and runtimeLoadMonitor?
- asim's addStoreToGossip was just missing the node capacity update, right? But I thought StoreDescriptors already populated desc.NodeCapacity? I looked at the asim code and it doesn't seem to be the case there, okay. I'm still a bit fuzzy on what was fixed here. It looks like the plumbing in production hasn't changed, so this should all be ok. (Sorry, very train of thought, but just to give you an idea of how much reverse engineering is going on in my head as I try to piece together three force push diffs).

> Things are already flaky based on CI: `require.GreaterOrEqual(t, nc.NodeCPURateCapacity, nc.NodeCPURateUsage)` failed with
> 
> ```
> pkg/testutils/soon.go:38
> pkg/kv/kvserver/store_test.go:4253
> Error:        "16000000000" is not greater than or equal to "24230879442"
> ```
> 
> I will try to stress it and see what is going on.


That's interesting. You probably did this already, but make sure it's in the spreadsheet too , we need to understand this better.

tbg commented:
bors r+

wenyihu6 commented:
> why did the test move to load_test?

I moved node_capacity_provider_test.go from the `load` package to `load_test` package since it only uses public methods like as an external user of the package.

> why is there an -ldl linker flag now?

There was a build failure due to a missing -ldl linker flag. From my chat with Cursor, this seems to be a long-running issue with `pkg/server/status` since it depends on `-ldl`, but this issue only surfaced now because previously, other packages in the dependency graph supplied the flag. And the load package doesn't supply it. 


> could you say a word about the plumbing changes? I see that NewStoreGossip used to take the provider, now it doesn't, at least based on [this force push](https://github.com/cockroachdb/cockroach/compare/a29c42e356ff701482c57e2569fc8ee53d7a6499..bb417f5149eba5d93c2e84896d5bc6b53f23204c)
That same diff also removes simNodeCapacityProvider, I'm assuming it was unused as a result of the previous merging of node capacity provider and runtimeLoadMonitor?
asim's addStoreToGossip was just missing the node capacity update, right? But I thought StoreDescriptors already populated desc.NodeCapacity? I looked at the asim code and it doesn't seem to be the case there, okay. I'm still a bit fuzzy on what was fixed here. It looks like the plumbing in production hasn't changed, so this should all be ok. (Sorry, very train of thought, but just to give you an idea of how much reverse engineering is going on in my head as I try to piece together three force push diffs).

 In our prototype, node capacity provider was passed into `StoreGossip` and the node capacity field of the store descriptor populated only after `DescriptorGetter.Descriptor` was called https://github.com/cockroachdb/cockroach/blob/4935635a00f5f92e0a1446d76a0a0449d84e9341/pkg/kv/kvserver/store_gossip.go#L384-L386. 

Master followed a similar structure by passing nil as the NodeCapacityProvider into NewStoreGossip and was left unused.

In the first commit, we removed the provider from kvserver.NewStoreGossip, and `NewStoreGossip` no longer takes a `NodeCapacityProvider`.

But the simulator was still using it. That force push updates the simulator to call into the state directly during descriptor getter https://github.com/cockroachdb/cockroach/blob/a2e2f92c764f2bb5b5cf556b266df753bdb06866/pkg/kv/kvserver/asim/gossip/gossip.go#L125-L130, instead of going through simNodeCapacityProvider, which was just wrapping that logic anyway https://github.com/cockroachdb/cockroach/blob/4935635a00f5f92e0a1446d76a0a0449d84e9341/pkg/kv/kvserver/asim/gossip/gossip.go#L130.

---

As I'm typing this, I realize how things became unclear and confusing as I didn’t add new commit messages to explain what changed in each force push, and Reviewable made it harder to follow. I share the sentiment and will try to provide a more pleasant review experience next time.
